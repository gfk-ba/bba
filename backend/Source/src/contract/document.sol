pragma solidity ^0.4.21;

contract Document {

    string    public data;
    address[] public links;

    constructor(string newData, address[] newLinks) public {
        data = newData;
        links = newLinks;
    }
}

// "", [0xE0f5206BBD039e7b0592d8918820024e2a7437b9]