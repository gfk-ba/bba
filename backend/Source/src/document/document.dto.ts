/**
 * The DTO (data transfer object) used to transport document data from the frontend to the backend.
 */
interface DocumentDto {
  /**
   * The id of the project to which the document belongs.
   * To be discussed: Why is this a string?
   */
  readonly projectId: string;

  /**
   * The kind of document (just a string)
   */
  readonly kind: string;

  /**
   * The hash of the document to store in the blockchain
   */
  readonly hash: string;

  /**
   * The description of the document (given by the user)
   */
  readonly description: string;

  /**
   * The list of linked documents (as a list of ids)
   */
  readonly linkIds: any;

  /**
   * The payment data (e.g. from PayPal)
   */
  readonly payment: any;
}
