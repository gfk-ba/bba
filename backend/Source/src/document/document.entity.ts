/**
 * This files contains the entity class for the document
 */

/**
 * import all decorator from typeorm
 */
import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm';

/**
 * The class Document is used with the ORM TypeORM to access data in the database table documents.
 */
@Entity('documents')
export class Document {
  /**
   * The ID (primary key) of the table (autogenerated)
   */
  @PrimaryGeneratedColumn()
  id: number;

  /**
   * The hashed data of the document
   */
  @Column({ type: 'varchar', length: 128 })
  hash: string;

  /**
   * The id of the project to which the document belongs (foreign key)
   */
  @Column({ type: 'int', name: 'project_id' })
  projectId: number;

  /**
   * The kind (type) of the document, must be choosen by the user in the frontend.
   */
  @Column({ type: 'varchar', length: 10 })
  kind: string;

  /**
   * The description of the document, can be provided by the user in the frontend.
   */
  @Column({ type: 'varchar', length: 255 })
  description: string;

  /**
   * The etherum address of the document
   */
  @Column({ type: 'varchar', length: 42, nullable: true })
  address: string;

  /**
   * The current state of the document (e.g. whether the document is saved the blockchain).
   * Could be OK, ERROR, PROGRESS...
   */
  @Column({ type: 'varchar', length: 10, default: 'PROGRESS' })
  state: string;

  /**
   * Details about the state of the document, especially error messages if something went wrong.
   */
  @Column({ type: 'varchar', length: 255, name: 'state_details', default: '' })
  stateDetails: string;

  /**
   * The timestamp when the document was stored in the database. \
   * Note: "timestamptz" is NOT a typo, but prevents SQL errors because timestamp is a reserved word.
   */
  @Column({ type: 'timestamptz', name: 'created_at', nullable: true })
  createdAt: Date;
}
