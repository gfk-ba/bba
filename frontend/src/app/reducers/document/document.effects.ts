import { Inject, Injectable } from '@angular/core';
import { APP_ENVIRONMENT } from '@dcs/ngx-tools';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { timer } from 'rxjs';
import { flatMap, mapTo, takeUntil } from 'rxjs/operators';
import { Environment } from '../../../environments/environment';
import {
  DocumentsListActionTypes,
  FetchDocumentsList,
} from './documents-list/documents-list.actions';

@Injectable()
export class DocumentEffects {
  @Effect()
  pollDocuments = this.actions$.pipe(
    ofType(DocumentsListActionTypes.StartPolling),
    flatMap(() =>
      timer(0, this.env.documentPollingTimer).pipe(
        takeUntil(
          this.actions$.pipe(
            ofType(DocumentsListActionTypes.StopPolling, DocumentsListActionTypes.StartPolling)
          )
        ),
        mapTo(new FetchDocumentsList())
      )
    )
  );

  constructor(private actions$: Actions, @Inject(APP_ENVIRONMENT) private env: Environment) {}
}
