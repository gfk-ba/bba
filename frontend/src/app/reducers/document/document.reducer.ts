import { ActionReducer, combineReducers } from '@ngrx/store';
import { ICurrentDocumentState } from './current-document/current-document.manager';
import { currentDocument } from './current-document/current-document.reducer';
import { DocumentsListActions } from './documents-list/documents-list.actions';
import { IDocumentsListState } from './documents-list/documents-list.manager';
import { documentsList } from './documents-list/documents-list.reducer';

export interface State {
  currentDocument: ICurrentDocumentState;
  documentsList: IDocumentsListState;
}

export const reducer: ActionReducer<State, DocumentsListActions> = combineReducers({
  currentDocument,
  documentsList,
});
