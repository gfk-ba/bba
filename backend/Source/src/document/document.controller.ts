/**
 * This files contains the DocumentController
 */

/**
 * import all the stuff we need from NestJS
 */
import {
  Controller,
  Get,
  Post,
  Body,
  UseGuards,
  Req,
  HttpException,
  HttpStatus,
} from '@nestjs/common';
import { DocumentService } from './document.service';
import { Document } from './document.entity';
import { AuthGuard } from '@nestjs/passport';
import { UserDataService } from '../shared/auth/user-data.service';
import * as logger from '../shared/logger';

/**
 * The DocumentController provides methods to find all documents of an user [[findAll]] and to create/store a new document [[createDocument]].
 * All methods are accessible at http://SERVER:PORT/documents
 */
@Controller('documents')
export class DocumentController {
  /**
   * Use dependency injection to get access to the [[DocumentService]] and the [[UserDataService]].
   */
  constructor(
    private readonly documentService: DocumentService,
    private userDataService: UserDataService,
  ) {}

  /**
   * This async function delivers the list of all documents of the user, who makes the call.
   * If user data is missing or the user is unknown, the result will the the http status code 417.
   * This function can be executed by a GET call to http://SERVER:PORT/documents
   * ```
   *   curl -X GET \
   *     http://SERVER:PORT/documents \
   *     -H 'Authorization: Bearer XXXXX' \
   *     -H 'Cache-Control: no-cache'
   * ```
   * This route is restricted by [[AuthGuard]]
   * @param requestData The request data, we need this to extract the user data
   * @returns The list of document of the current user (including links to other documents).
   */
  @Get()
  @UseGuards(AuthGuard())
  async findAll(@Req() requestData): Promise<Document[]> {
    const userId = this.userDataService.getUserData(requestData);
    if (!userId) {
      throw new HttpException(
        'userId missing in DocumentController.findAll',
        HttpStatus.EXPECTATION_FAILED,
      );
    }

    const result = await this.documentService.findAll(userId);
    return result;
  }
  /**
   * This async function creates+stores a new document using the request data. The document will be associated to the user, who makes the call.
   * If user data is missing or the user is unknown, the result will the the http status code 417.
   * This function can be executed by a POST call to http://SERVER:PORT/documents
   * ```
   * curl -X POST \
   *    http://SERVER:PORT/documents \
   *   -H 'Authorization: Bearer XXXXX' \
   *   -H 'Cache-Control: no-cache' \
   *   -H 'Content-Type: application/x-www-form-urlencoded' \
   *   -d 'projectId=3&kind=result&hash=12121212&description=testdaten&linkIds=%5B2%2C%204%5D'
   * ```
   * The result should be something like this:
   * ```
   * {
   *        "projectId": 3,
   *        "kind": "result",
   *        "hash": "12121212",
   *        "description": "testdaten",
   *        "address": null,
   *        "id": 4,
   *        "state": "PROGRESS",
   *        "stateDetails": "",
   *        "createdAt": "2018-09-16"
   * }
   * ```
   * This route is restricted by [[AuthGuard]]
   * @param documentDto The data of the new document
   * @param requestData The request data, we need this to extract the user data
   * @returns the newly created document - including the database id.
   */
  @Post()
  @UseGuards(AuthGuard())
  async createDocument(
    @Body() documentDto: DocumentDto,
    @Req() requestData,
  ): Promise<any> {
    logger.log('DocumentController.createDocument - documentDto', documentDto);

    // documentDto should be something like this ....
    // {projectId: "3",
    // kind: "result",
    // hash: "212121",
    // description: "ttttt",
    // linkIds: [2, 4]}
    // payment: { }

    if (!this.documentService.hasValidPaymentData) {
      throw new HttpException(
        'payment receipt is missing',
        HttpStatus.EXPECTATION_FAILED,
      );
    }

    const userId = this.userDataService.getUserData(requestData);

    const savedDocument = await this.documentService.createDocument(
      userId,
      documentDto,
    );

    logger.log(
      'DocumentController.createDocument - savedDocument',
      savedDocument,
    );

    return savedDocument;
  }
}
