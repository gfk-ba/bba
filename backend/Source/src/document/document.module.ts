/**
 * This file contains the [[DocumentModule]] which provides the [[DocumentService]] and [[DocumentController]] to other modules
 */

/**
 * Importing stuff from NestJS and our own code.
 */
import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { DocumentService } from './document.service';
import { DocumentController } from './document.controller';
import { Document } from './document.entity';
import { PaymentModule } from '../payment/payment.module';
import { ContractModule } from '../contract/contract.module';
import { AuthModule } from '../shared/auth/auth.module';
import { ProjectModule } from '../project/project.module';
import { LinkModule } from '../link/link.module';

/**
 * This module provides the [[DocumentService]] and [[DocumentController]] to other modules
 */
@Module({
  imports: [
    TypeOrmModule.forFeature([Document]),
    AuthModule,
    ProjectModule,
    LinkModule,
    AuthModule,
    PaymentModule,
    ContractModule,
  ],
  providers: [DocumentService],
  controllers: [DocumentController],
  exports: [DocumentService],
})
export class DocumentModule {}
