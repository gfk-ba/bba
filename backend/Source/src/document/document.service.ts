/**
 * Contains the [[DocumentService]]
 */

/**
 * Import NestJS and TypeORM stuff
 */
import { Injectable, HttpException, HttpStatus } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';

import { Document } from './document.entity';
import { LinkService } from '../link/link.service';
import { PaymentService } from '../payment/payment.service';

import * as logger from '../shared/logger';
import { ContractService } from '../contract/contract.service';
import { ProjectService } from '../project/project.service';
import { Link } from '../link/link.entity';

/**
 * The DocumentService provides a lot of functions to find documents, their links etc.
 */
@Injectable()
export class DocumentService {
  /**
   * Use dependency injection to get access lots of services
   * @param documentRepository the repository for documents (used to access the database table documents)
   * @param projectService the service for the projects (e.g. lookup projects)
   * @param linkService the service for links  (e.g. lookup liks to a document)
   * @param paymentService the payment service (e.g. store payment data - either coupons or PayPal)
   * @param contractService the contract service (e.g. create a contract on the blockchain for the document)
   */
  constructor(
    @InjectRepository(Document)
    private readonly documentRepository: Repository<Document>,
    private projectService: ProjectService,
    private linkService: LinkService,
    private paymentService: PaymentService,
    private contractService: ContractService,
  ) {}

  /**
   * This async function retrieves the list of all documents (including links) for the given user id
   * @param userId The id of the user for whom we want the list of all documents.
   * @returns The list of all documents (including links) or http 417 if the user id is missing/empty.
   */
  async findAll(userId: number): Promise<any[]> {
    if (!userId) {
      throw new HttpException(
        'userId missing in DocumentService.findAll',
        HttpStatus.EXPECTATION_FAILED,
      );
    }

    const result = await this.findAllForUserId(userId);
    const linksGroupedBySource = await this.linkService.findAll();

    const resultList = result.map((doc: any) =>
      this.addLinks(doc, linksGroupedBySource),
    );

    return resultList;
  }

  /**
   * This async function retrieves a list of documents from the given list of document ids.
   * @param listOfDocIds The list of document ids.
   * @returns The list of the documents.
   */
  async findAllForDocIds(
    listOfDocIds: Array<number>,
  ): Promise<Array<Document>> {
    if (!listOfDocIds || !listOfDocIds.length) {
      return [];
    }

    const result = await this.documentRepository.manager
      .createQueryBuilder(Document, 'document')
      .where('document.id IN (:...listOfDocIds)', {
        listOfDocIds,
      })
      .printSql()
      .getMany();

    return result;
  }

  /**
   * This async function retrieves the list of all linked documents of the given document id
   * @param sourceDocumentId The id of the document for which we want all linked documents
   * @returns The list of linked documents.
   */
  async findAllLinkedDocs(sourceDocumentId: number): Promise<Array<Document>> {
    logger.log('findAllLinkedDocs - sourceDocumentId', sourceDocumentId);

    // Get all the links from the source document
    const allLinks =
      (await this.linkService.findAllLinksForDocument(sourceDocumentId)) || [];

    // .. get the ids of the target documents
    const allTargetDocIds = allLinks.map((link: Link) => link.targetDocumentId);

    // .. and get the document for each (target document) id
    const result = await this.findAllForDocIds(allTargetDocIds);

    logger.log('findAllLinkedDocs - result', result);

    return result;
  }

  /**
   * Add all links (more precise: a list of the ids for all linked documents) to document.
   * @param doc The document to which we want to add the links.
   * @param linksGroupedBySource A data structure which contains a hash with source document id as key and an array of target document ids as value.
   */
  private addLinks(doc: Document, linksGroupedBySource: any) {
    const result: any = doc;
    result.linkIds = linksGroupedBySource[result.id] || [];
    return result;
  }

  /**
   * This private async helper function retrieves the list of all document for the given user (specified by userId).
   * @param userId The id of the user for whom we want get the list of documents.
   * @returns The list of all documents (without links) for the given user (specified by userId)
   */
  private async findAllForUserId(userId: number): Promise<Array<Document>> {
    logger.log(`DocumentService.findAllForUserId(${userId})`);

    const projects = await this.projectService.findAll(userId);

    if (projects && projects.length) {
      const projectIds = projects.map(p => p.id);

      const result = await this.documentRepository.manager
        .createQueryBuilder(Document, 'document')
        .where('document.project_id IN (:...projects)', {
          projects: projectIds,
        })
        .printSql()
        .getMany();

      return result;
    }

    return [];
  }

  /**
   * This async function creates a new document for the given user  (specified by userId).
   * @param userId The id of the user for whom we want to create a new document.
   * @param documentDto The data of the document we should create.
   * @returns the new document (including state information)
   */
  async createDocument(
    userId: number,
    documentDto: DocumentDto,
  ): Promise<Document> {
    logger.log(`DocumentService.createDocument(${userId}, ${documentDto})`);

    // projectId shoudl contains something like this ...
    // {projectId: "3",
    // kind: "result",
    // hash: "212121",
    // description: "ttttt",
    // linkIds: [2, 4]}

    if (!userId) {
      throw new HttpException(
        'userId missing in DocumentService.createDocument',
        HttpStatus.EXPECTATION_FAILED,
      );
    }

    // We could improve security here by double-checking that the project (specified by documentDto.projectId)
    // belongs to the user (specified by userId)

    const NEW_DOCUMENT = new Document();
    NEW_DOCUMENT.projectId = parseInt(documentDto.projectId, 10);
    NEW_DOCUMENT.kind = documentDto.kind;
    NEW_DOCUMENT.hash = documentDto.hash;
    NEW_DOCUMENT.description = documentDto.description;
    NEW_DOCUMENT.createdAt = new Date();

    logger.log('DocumentService.createDocument - doc', NEW_DOCUMENT);

    const SAVED_DOCUMENT = await this.documentRepository.save(NEW_DOCUMENT);

    logger.log(
      'DocumentService.createDocument - savedDocument',
      SAVED_DOCUMENT,
    );

    const result = await this.findById(SAVED_DOCUMENT.id);
    logger.log('DocumentService.createDocument - result', result);

    await this.storeLinksToDoc(SAVED_DOCUMENT.id, documentDto);
    await this.processPaymentData(SAVED_DOCUMENT.id, documentDto.payment || {});

    // Alternative solution without await:
    // this.addToBlockchain(savedDocument, documentDto).then(address => {
    //   this.documentRepository.update(savedDocument.id, {
    //     address,
    //     state: 'OK',
    //     stateDetails:
    //       'Successfully saved in the ethereum blockchain at ' + new Date(),
    //   });
    // });

    const address = await this.addToBlockchain(SAVED_DOCUMENT, documentDto);

    logger.log(
      `DocumentService.createDocument - blockchain address = ${address}`,
    );

    SAVED_DOCUMENT.address = address;
    SAVED_DOCUMENT.state = 'OK';
    SAVED_DOCUMENT.stateDetails =
      'Successfully saved in the ethereum blockchain at ' + new Date();

    const DOC_FROM_DATABASE = await this.documentRepository.save(
      SAVED_DOCUMENT,
    );

    logger.log(
      `DocumentService.createDocument - nach addToBlockchain -  DOC_FROM_DATABASE = `,
      DOC_FROM_DATABASE,
    );

    return SAVED_DOCUMENT;
  }

  /**
   * This private async helper function stores a document in the blockchain.
   * @param savedDocument the document that should be stored in the blockchain
   * @param documentDto The document DTO, we use this to get the blockchain addresses of the linked documents
   * @returns the blockchain address of the document
   */
  private async addToBlockchain(
    savedDocument: Document,
    documentDto: DocumentDto,
  ): Promise<string> {
    const addressesOfLinkedDocs = await this.getAdressesOfLinksToDoc(
      documentDto,
    );

    return await this.contractService.createContract(
      savedDocument.hash,
      addressesOfLinkedDocs,
    );
  }

  /**
   * This private async helper function creates links (in the database) from the document
   * (specified by sourceDocumentId) to the linked documents (stored in the documentDto)
   * @param sourceDocumentId the id of the document that's the source of a created links
   * @param documentDto The document DTO which already contains a list of target ids
   */
  private async storeLinksToDoc(
    sourceDocumentId: number,
    documentDto: DocumentDto,
  ) {
    logger.log(
      'DocumentService.storeLinksToDoc -  documentDto.linkIds',
      documentDto.linkIds,
    );

    const linkIds = this.getIdsOfLinkedDocuments(documentDto);

    linkIds.forEach(async targetId => {
      logger.log('DocumentService.storeLinksToDoc - targetId', targetId);
      await this.linkService.create(sourceDocumentId, targetId);
    });
  }

  /**
   * This private helper function retrieves of list of id belonging to the target documents for the given document
   * (specified by documentDto)
   * @param documentDto The document DTO from which we get the target documents ids
   * @returns The list of ids belonging to the documents linked to the documentDto
   */
  private getIdsOfLinkedDocuments(documentDto: DocumentDto): Array<number> {
    let linkIds = [];

    if (
      documentDto &&
      documentDto.linkIds &&
      Array.isArray(documentDto.linkIds)
    ) {
      linkIds = documentDto.linkIds;
    }

    logger.log('DocumentService.getIdsOfLinkedDocuments - linkIds', linkIds);

    return linkIds;
  }

  /**
   * This private async helper function retrieves of list of blockchain addresses belonging to the target documents for the given document
   * (specified by documentDto)
   * @param documentDto The document DTO from which we get the target documents ids
   * @returns The list of blockchain addresses belonging to the documents linked to the documentDto
   */
  private async getAdressesOfLinksToDoc(
    documentDto: DocumentDto,
  ): Promise<Array<string>> {
    const linkIds = this.getIdsOfLinkedDocuments(documentDto);
    const linkedDocs = await this.findAllForDocIds(linkIds);
    const result = linkedDocs.map(document => document.address || '');

    logger.log('DocumentService.getAdressesOfLinksToDoc - result', result);

    return result;
  }

  /**
   * This private async helper function will double check and store the payment data
   * @param documentId The id of document from which we check+save the payment data.
   * @param paymentData The payment data. The structure depends on the payment method (PayPal or coupon)
   * @returns true, if the payment data is consistent and can be saved, false if not.
   */
  private async processPaymentData(documentId: number, paymentData: any) {
    logger.log('DocumentService.processPaymentData - started', {
      documentId,
      paymentData,
    });

    if (!paymentData.coupon && !paymentData.paypal) {
      logger.log(
        'DocumentService.processPaymentData - ERROR no valid payment method found',
        {
          documentId,
          paymentData,
        },
      );
      return false;
    }

    if (paymentData.coupon && paymentData.paypal) {
      logger.log(
        'DocumentService.processPaymentData - ERROR only one payment method may be provided',
        { documentId, paymentData },
      );
      return false;
    }

    logger.log(`DocumentService.processPaymentData - paymentData`, paymentData);

    let payment;

    if (paymentData.coupon) {
      payment = await this.paymentService.storeCouponPayment(
        documentId,
        paymentData.coupon,
      );
    } else {
      payment = await this.paymentService.storePayPalPayment(
        documentId,
        paymentData.paypal,
      );
    }

    logger.log('DocumentService.processPaymentData - payment record', payment);

    return true;
  }

  /**
   * This function will double check the payment data
   * @param documentDto The document DTO from which we get the payment data (at least we hope ...)
   * @returns true, if the payment data is consistent, false if not.
   */
  hasValidPaymentData(documentDto: DocumentDto): boolean {
    // Note: There's some overlapping functionality with processPaymentData
    // Refactoring could reduce duplicate code

    const paymentData = documentDto.payment;

    if (!paymentData) {
      logger.log(
        'DocumentService.hasValidPaymentData - ERROR no payment data found',
        {
          documentDto,
        },
      );
      return false;
    }

    let result = true;

    if (!paymentData.coupon && !paymentData.paypal) {
      logger.log(
        'DocumentService.hasValidPaymentData - ERROR no valid payment method found',
        {
          paymentData,
        },
      );
      result = false;
    }

    if (paymentData.coupon && paymentData.paypal) {
      logger.log(
        'DocumentService.hasValidPaymentData - ERROR only one payment method may be provided',
        { paymentData },
      );
      result = false;
    }

    return result;
  }

  /**
   * This private async helper function will load the document (including all links) specified by the document id
   * @param documentId The id of document which we want to retrieve (including all links).
   * @returns The document (including all links).
   */
  private async findById(documentID: number): Promise<any> {
    const linksGroupedBySource = await this.linkService.findAll();
    const doc = await this.documentRepository.findOne({ id: documentID });
    const result = this.addLinks(doc, linksGroupedBySource);
    return result;
  }
}
