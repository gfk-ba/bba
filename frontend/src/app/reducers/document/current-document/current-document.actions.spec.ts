import { Document } from '../models/document.class';
import {
  CreateCurrentDocument,
  FetchCurrentDocument,
  UpdateCurrentDocument,
  DeleteCurrentDocument,
  ResetCurrentDocument,
} from './current-document.actions';

describe('current document actions', () => {
  describe('FetchCurrentDocument', () => {
    it('builds the correct API request action', () => {
      expect(new FetchCurrentDocument('12')).toMatchSnapshot();
    });
  });

  describe('CreateCurrentDocument', () => {
    it('builds the correct API request action', () => {
      expect(
        new CreateCurrentDocument({
          id: '',
          projectId: 0,
          description: '',
          kind: 'data',
          hash: '',
          linkIds: [],
          links: [],
          project: null,
          state: 'in_progress',
          stateDetails: '',
          address: '',
          createdAt: '',
        })
      ).toMatchSnapshot();
    });
  });

  describe('UpdateCurrentDocument', () => {
    it('builds the correct API request action', () => {
      expect(new UpdateCurrentDocument(new Document())).toMatchSnapshot();
    });
  });

  describe('DeleteCurrentDocument', () => {
    it('builds the correct API request action', () => {
      expect(new DeleteCurrentDocument(new Document({ id: '12' }))).toMatchSnapshot();
    });
  });

  describe('ResetCurrentDocument', () => {
    it('builds the correct API request action', () => {
      expect(new ResetCurrentDocument()).toMatchSnapshot();
    });
  });
});
