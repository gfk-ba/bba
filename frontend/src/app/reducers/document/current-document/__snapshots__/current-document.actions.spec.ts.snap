// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`current document actions CreateCurrentDocument builds the correct API request action 1`] = `
CreateCurrentDocument {
  "payload": Object {
    "cancel": undefined,
    "handlers": "[Current Document] Create",
    "normalizrSchema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "documents",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "schema": Object {
        "links": Array [
          EntitySchema {
            "_getId": [Function],
            "_idAttribute": "id",
            "_key": "documents",
            "_mergeStrategy": [Function],
            "_processStrategy": [Function],
            "schema": Object {},
          },
        ],
        "project": EntitySchema {
          "_getId": [Function],
          "_idAttribute": [Function],
          "_key": "projects",
          "_mergeStrategy": [Function],
          "_processStrategy": [Function],
          "schema": Object {},
        },
      },
    },
    "rawDataProcessor": undefined,
    "request": Object {
      "method": "POST",
      "options": Object {
        "body": Object {
          "address": "",
          "createdAt": "",
          "description": "",
          "hash": "",
          "id": "",
          "kind": "data",
          "linkIds": Array [],
          "links": Array [],
          "project": null,
          "projectId": 0,
          "state": "in_progress",
          "stateDetails": "",
        },
      },
      "url": "documents",
    },
  },
  "type": "[ApiRequest] Execute API Request",
}
`;

exports[`current document actions DeleteCurrentDocument builds the correct API request action 1`] = `
DeleteCurrentDocument {
  "payload": Object {
    "cancel": undefined,
    "handlers": "[Current Document] Delete",
    "normalizrSchema": undefined,
    "rawDataProcessor": undefined,
    "request": Object {
      "method": "DELETE",
      "options": Object {
        "body": Object {
          "address": "",
          "createdAt": "",
          "description": "",
          "hash": "",
          "id": "12",
          "kind": "data",
          "linkIds": Array [],
          "links": Array [],
          "project": null,
          "projectId": 0,
          "state": "in_progress",
          "stateDetails": "",
        },
      },
      "url": "documents/12",
    },
  },
  "type": "[ApiRequest] Execute API Request",
}
`;

exports[`current document actions FetchCurrentDocument builds the correct API request action 1`] = `
FetchCurrentDocument {
  "payload": Object {
    "cancel": undefined,
    "handlers": "[Current Document] Fetch",
    "normalizrSchema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "documents",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "schema": Object {
        "links": Array [
          EntitySchema {
            "_getId": [Function],
            "_idAttribute": "id",
            "_key": "documents",
            "_mergeStrategy": [Function],
            "_processStrategy": [Function],
            "schema": Object {},
          },
        ],
        "project": EntitySchema {
          "_getId": [Function],
          "_idAttribute": [Function],
          "_key": "projects",
          "_mergeStrategy": [Function],
          "_processStrategy": [Function],
          "schema": Object {},
        },
      },
    },
    "rawDataProcessor": undefined,
    "request": Object {
      "method": "GET",
      "url": "documents/12",
    },
  },
  "type": "[ApiRequest] Execute API Request",
}
`;

exports[`current document actions ResetCurrentDocument builds the correct API request action 1`] = `
ResetCurrentDocument {
  "type": "[Current Document] Fetch RESET",
}
`;

exports[`current document actions UpdateCurrentDocument builds the correct API request action 1`] = `
UpdateCurrentDocument {
  "payload": Object {
    "cancel": undefined,
    "handlers": "[Current Document] Update",
    "normalizrSchema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "documents",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "schema": Object {
        "links": Array [
          EntitySchema {
            "_getId": [Function],
            "_idAttribute": "id",
            "_key": "documents",
            "_mergeStrategy": [Function],
            "_processStrategy": [Function],
            "schema": Object {},
          },
        ],
        "project": EntitySchema {
          "_getId": [Function],
          "_idAttribute": [Function],
          "_key": "projects",
          "_mergeStrategy": [Function],
          "_processStrategy": [Function],
          "schema": Object {},
        },
      },
    },
    "rawDataProcessor": undefined,
    "request": Object {
      "method": "PUT",
      "options": Object {
        "body": Object {
          "address": "",
          "createdAt": "",
          "description": "",
          "hash": "",
          "id": "",
          "kind": "data",
          "linkIds": Array [],
          "links": Array [],
          "project": null,
          "projectId": 0,
          "state": "in_progress",
          "stateDetails": "",
        },
      },
      "url": "documents/",
    },
  },
  "type": "[ApiRequest] Execute API Request",
}
`;
